* data science is the extention of data analyst(he find the pattern in data)
* but data scientist do predictive modeling
* [A data set](https://www.kaggle.com/lava18/google-play-store-apps/home) containing data about approximately ten thousand Android apps from Google Play
* understand the user data is called data dictionary which we perform withthe     help of client 



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

android=pd.read_csv("googleplaystore.csv",names=["1","2","3","4","5","6","7","8","9","10"])
android

In [2]:
android=pd.read_csv("googleplaystore.csv")
android

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

In [3]:
android.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
android.tail()

App             Category  \
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10836       Everyone          Education     July 25, 2017                1.48   
10837       Everyone          Education      July 6, 2018                 1.0   
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [5]:
type(android)

pandas.core.frame.DataFrame

In [6]:
android.shape

(10841, 13)

In [7]:
10841*13 #total data points

140933

In [8]:
android.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

In [10]:
android.describe(include="all")

App Category       Rating Reviews                Size    Installs  \
count    10841    10841  9367.000000   10841               10841       10841   
unique    9660       34          NaN    6002                 462          22   
top     ROBLOX   FAMILY          NaN       0  Varies with device  1,000,000+   
freq         9     1972          NaN     596                1695        1579   
mean       NaN      NaN     4.193338     NaN                 NaN         NaN   
std        NaN      NaN     0.537431     NaN                 NaN         NaN   
min        NaN      NaN     1.000000     NaN                 NaN         NaN   
25%        NaN      NaN     4.000000     NaN                 NaN         NaN   
50%        NaN      NaN     4.300000     NaN                 NaN         NaN   
75%        NaN      NaN     4.500000     NaN                 NaN         NaN   
max        NaN      NaN    19.000000     NaN                 NaN         NaN   

         Type  Price Content Rating Genres    Last Updated  \
count   10840  10841          10840  10841           10841   
unique      3     93              6    120            1378   
top      Free      0       Everyone  Tools  August 3, 2018   
freq    10039  10040           8714    842             326   
mean      NaN    NaN            NaN    NaN             NaN   
std       NaN    NaN            NaN    NaN             NaN   
min       NaN    NaN            NaN    NaN             NaN   
25%       NaN    NaN            NaN    NaN             NaN   
50%       NaN    NaN            NaN    NaN             NaN   
75%       NaN    NaN            NaN    NaN             NaN   
max       NaN    NaN            NaN    NaN             NaN   

               Current Ver Android Ver  
count                10833       10838  
unique                2832          33  
top     Varies with device  4.1 and up  
freq                  1459        2451  
mean                   NaN         NaN  
std                    NaN         NaN  
min                    NaN         NaN  
25%                    NaN         NaN  
50%                    NaN         NaN  
75%                    NaN         NaN  
max                    NaN         NaN

In [9]:
android.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [11]:
android.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [12]:
android.index

RangeIndex(start=0, stop=10841, step=1)

In [13]:
len(android)

10841

In [14]:
android.ndim

2

In [23]:
android.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

In [24]:
len(pd.notnull(android))

10841

In [25]:
a=len(android["Rating"])
a

10841

In [26]:
b=len(android[pd.notnull(android["Rating"])])
b

9367

In [27]:
b=len(android[pd.isnull(android["Rating"])])
b

1474

In [28]:
a-b  # missing value in ratonng columns

9367

In [29]:
android.loc[10472] # data one column move left

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              1.9
Rating                                                 19
Reviews                                              3.0M
Size                                               1,000+
Installs                                             Free
Type                                                    0
Price                                            Everyone
Content Rating                                        NaN
Genres                                  February 11, 2018
Last Updated                                       1.0.19
Current Ver                                    4.0 and up
Android Ver                                           NaN
Name: 10472, dtype: object

In [30]:
android.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              1.9
Rating                                                 19
Reviews                                              3.0M
Size                                               1,000+
Installs                                             Free
Type                                                    0
Price                                            Everyone
Content Rating                                        NaN
Genres                                  February 11, 2018
Last Updated                                       1.0.19
Current Ver                                    4.0 and up
Android Ver                                           NaN
Name: 10472, dtype: object

In [31]:
android.head(1)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   

  Reviews Size Installs  Type Price Content Rating        Genres  \
0     159  19M  10,000+  Free     0       Everyone  Art & Design   

      Last Updated Current Ver   Android Ver  
0  January 7, 2018       1.0.0  4.0.3 and up

**4 way to handle missing data**
* remove row
* column remove
* fill missing data, using mean , median and mode
* remain as it is

In [32]:
android.drop(10472,axis=0,inplace=True) #by default remove row

In [33]:
android

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10840 rows x 13 columns]

In [36]:
android.iloc[10472] #back end index

App               osmino Wi-Fi: free WiFi
Category                            TOOLS
Rating                                4.2
Reviews                            134203
Size                                 4.1M
Installs                      10,000,000+
Type                                 Free
Price                                   0
Content Rating                   Everyone
Genres                              Tools
Last Updated               August 7, 2018
Current Ver                       6.06.14
Android Ver                    4.4 and up
Name: 10473, dtype: object

**Name: 10473, dtype: object in backend order will remain same but front end index will update**
* iloc means give the index which is in backend
* lock mean give the index which is in front end
* when we use drop any row it use front end (label indexing) to remove that row

In [37]:
android.loc[10472] # front end index , there for we got error here 

KeyError: 10472

In [38]:
android.loc[10473] # front end label indexing

App               osmino Wi-Fi: free WiFi
Category                            TOOLS
Rating                                4.2
Reviews                            134203
Size                                 4.1M
Installs                      10,000,000+
Type                                 Free
Price                                   0
Content Rating                   Everyone
Genres                              Tools
Last Updated               August 7, 2018
Current Ver                       6.06.14
Android Ver                    4.4 and up
Name: 10473, dtype: object

In [39]:
android.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9366 non-null   float64
 3   Reviews         10840 non-null  object 
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.5+ MB


In [40]:
android.duplicated(["App"])

0        False
1        False
2        False
3        False
4        False
         ...  
10836    False
10837    False
10838    False
10839    False
10840    False
Length: 10840, dtype: bool

In [41]:
android.duplicated(["App"]).sum() # mean these apps are duplicate

1181

In [42]:
10840-1181  # mean these apps are not duplicate

9659

 * facebook false
 * facebook true
 * facebook true
 * 0+1+1

In [43]:
bol=android.duplicated(["App"])
bol #will show only true result, this is bolean indexing

0        False
1        False
2        False
3        False
4        False
         ...  
10836    False
10837    False
10838    False
10839    False
10840    False
Length: 10840, dtype: bool

In [44]:
#we can use this bol as index, there is no need of loc and iloc but we can use as well iloc and loc, belo is duplicated value 
#and true value
android[bol]

App       Category  Rating Reviews  \
229          Quick PDF Scanner + OCR FREE       BUSINESS     4.2   80805   
236                                   Box       BUSINESS     4.2  159872   
239                    Google My Business       BUSINESS     4.4   70991   
256                   ZOOM Cloud Meetings       BUSINESS     4.4   31614   
261             join.me - Simple Meetings       BUSINESS     4.0    6989   
...                                   ...            ...     ...     ...   
10715                  FarmersOnly Dating         DATING     3.0    1145   
10720  Firefox Focus: The privacy browser  COMMUNICATION     4.4   36981   
10730                         FP Notebook        MEDICAL     4.5     410   
10753      Slickdeals: Coupons & Shopping       SHOPPING     4.5   33599   
10768                                AAFP        MEDICAL     3.8      63   

                     Size     Installs  Type Price Content Rating  \
229    Varies with device   5,000,000+  Free     0       Everyone   
236    Varies with device  10,000,000+  Free     0       Everyone   
239    Varies with device   5,000,000+  Free     0       Everyone   
256                   37M  10,000,000+  Free     0       Everyone   
261    Varies with device   1,000,000+  Free     0       Everyone   
...                   ...          ...   ...   ...            ...   
10715                1.4M     100,000+  Free     0     Mature 17+   
10720                4.0M   1,000,000+  Free     0       Everyone   
10730                 60M      50,000+  Free     0       Everyone   
10753                 12M   1,000,000+  Free     0       Everyone   
10768                 24M      10,000+  Free     0       Everyone   

              Genres       Last Updated         Current Ver  \
229         Business  February 26, 2018  Varies with device   
236         Business      July 31, 2018  Varies with device   
239         Business      July 24, 2018    2.19.0.204537701   
256         Business      July 20, 2018      4.1.28165.0716   
261         Business      July 16, 2018           4.3.0.508   
...              ...                ...                 ...   
10715         Dating  February 25, 2016                 2.2   
10720  Communication       July 6, 2018                 5.2   
10730        Medical     March 24, 2018           2.1.0.372   
10753       Shopping      July 30, 2018                 3.9   
10768        Medical      June 22, 2018               2.3.1   

              Android Ver  
229          4.0.3 and up  
236    Varies with device  
239            4.4 and up  
256            4.0 and up  
261            4.4 and up  
...                   ...  
10715          4.0 and up  
10720          5.0 and up  
10730          4.4 and up  
10753          4.4 and up  
10768          5.0 and up  

[1181 rows x 13 columns]

In [45]:
android.loc[bol]

App       Category  Rating Reviews  \
229          Quick PDF Scanner + OCR FREE       BUSINESS     4.2   80805   
236                                   Box       BUSINESS     4.2  159872   
239                    Google My Business       BUSINESS     4.4   70991   
256                   ZOOM Cloud Meetings       BUSINESS     4.4   31614   
261             join.me - Simple Meetings       BUSINESS     4.0    6989   
...                                   ...            ...     ...     ...   
10715                  FarmersOnly Dating         DATING     3.0    1145   
10720  Firefox Focus: The privacy browser  COMMUNICATION     4.4   36981   
10730                         FP Notebook        MEDICAL     4.5     410   
10753      Slickdeals: Coupons & Shopping       SHOPPING     4.5   33599   
10768                                AAFP        MEDICAL     3.8      63   

                     Size     Installs  Type Price Content Rating  \
229    Varies with device   5,000,000+  Free     0       Everyone   
236    Varies with device  10,000,000+  Free     0       Everyone   
239    Varies with device   5,000,000+  Free     0       Everyone   
256                   37M  10,000,000+  Free     0       Everyone   
261    Varies with device   1,000,000+  Free     0       Everyone   
...                   ...          ...   ...   ...            ...   
10715                1.4M     100,000+  Free     0     Mature 17+   
10720                4.0M   1,000,000+  Free     0       Everyone   
10730                 60M      50,000+  Free     0       Everyone   
10753                 12M   1,000,000+  Free     0       Everyone   
10768                 24M      10,000+  Free     0       Everyone   

              Genres       Last Updated         Current Ver  \
229         Business  February 26, 2018  Varies with device   
236         Business      July 31, 2018  Varies with device   
239         Business      July 24, 2018    2.19.0.204537701   
256         Business      July 20, 2018      4.1.28165.0716   
261         Business      July 16, 2018           4.3.0.508   
...              ...                ...                 ...   
10715         Dating  February 25, 2016                 2.2   
10720  Communication       July 6, 2018                 5.2   
10730        Medical     March 24, 2018           2.1.0.372   
10753       Shopping      July 30, 2018                 3.9   
10768        Medical      June 22, 2018               2.3.1   

              Android Ver  
229          4.0.3 and up  
236    Varies with device  
239            4.4 and up  
256            4.0 and up  
261            4.4 and up  
...                   ...  
10715          4.0 and up  
10720          5.0 and up  
10730          4.4 and up  
10753          4.4 and up  
10768          5.0 and up  

[1181 rows x 13 columns]

In [46]:
a=android["App"]=="Box"
a.sum()

3

In [47]:
android[android["App"]=="Box"]

App  Category  Rating Reviews                Size     Installs  Type  \
204  Box  BUSINESS     4.2  159872  Varies with device  10,000,000+  Free   
236  Box  BUSINESS     4.2  159872  Varies with device  10,000,000+  Free   
265  Box  BUSINESS     4.2  159872  Varies with device  10,000,000+  Free   

    Price Content Rating    Genres   Last Updated         Current Ver  \
204     0       Everyone  Business  July 31, 2018  Varies with device   
236     0       Everyone  Business  July 31, 2018  Varies with device   
265     0       Everyone  Business  July 31, 2018  Varies with device   

            Android Ver  
204  Varies with device  
236  Varies with device  
265  Varies with device

In [48]:
android[android["App"]=="Instagram"] # we will keep max review value data

App Category  Rating   Reviews                Size  \
2545  Instagram   SOCIAL     4.5  66577313  Varies with device   
2604  Instagram   SOCIAL     4.5  66577446  Varies with device   
2611  Instagram   SOCIAL     4.5  66577313  Varies with device   
3909  Instagram   SOCIAL     4.5  66509917  Varies with device   

            Installs  Type Price Content Rating  Genres   Last Updated  \
2545  1,000,000,000+  Free     0           Teen  Social  July 31, 2018   
2604  1,000,000,000+  Free     0           Teen  Social  July 31, 2018   
2611  1,000,000,000+  Free     0           Teen  Social  July 31, 2018   
3909  1,000,000,000+  Free     0           Teen  Social  July 31, 2018   

             Current Ver         Android Ver  
2545  Varies with device  Varies with device  
2604  Varies with device  Varies with device  
2611  Varies with device  Varies with device  
3909  Varies with device  Varies with device

In [49]:
android.Reviews

0           159
1           967
2         87510
3        215644
4           967
          ...  
10836        38
10837         4
10838         3
10839       114
10840    398307
Name: Reviews, Length: 10840, dtype: object

In [50]:
android["Reviews"]

0           159
1           967
2         87510
3        215644
4           967
          ...  
10836        38
10837         4
10838         3
10839       114
10840    398307
Name: Reviews, Length: 10840, dtype: object

In [51]:
type(android["Reviews"])

pandas.core.series.Series

In [52]:
android.info() #view has object data type

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9366 non-null   float64
 3   Reviews         10840 non-null  object 
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.5+ MB


* 3 main data type
number is on top, which have 2 braches 
* intiger (whole number)
* float (decimal or fraction)
* object (in python we we call object as string, no mathatical equation can we perform )

In [53]:
android["Reviews"].sum() #this is not a number datatype, so we need to convert it into int type

'159967875102156449671671783681513791121138808788448294326151855363227194216224399450654769961118192202602031362231120227503510153535648145366391585911171762952212206261745311070858453671598284170571295421047980512914033971534777438846243160902951904021135652530116986137927192073670211977373574994197136142151682155138541421777348250133727880361748066578631433509717542680128818900497901150173932090222543698572964420501041793460136187182301347411371535959315753826834119227722806541849773536418473306638712576229147241857447857755081408024631545406015544641822486203130143511650714332339046886036393496787873175061862208447303190801618584278319161546202133626875177827096451383425271322168027391065233757287885126930105156114341157161296747301254730851853258421773070991100286116589148945445862272894146353127918488073670001598723084718884111622959124034459641495569033161423055190232073721225380837106007435982280805228741621476023243989788080542492286897103755465051144210295159872296293137099151

In [54]:
android["Reviews"]=android["Reviews"].astype(float)

In [55]:
android.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9366 non-null   float64
 3   Reviews         10840 non-null  float64
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.5+ MB


In [56]:
android["Reviews"].head()

0       159.0
1       967.0
2     87510.0
3    215644.0
4       967.0
Name: Reviews, dtype: float64

In [57]:
android.sort_values("Reviews",ascending=False).head(15) #data sort with respect to max review is at top

App       Category  Rating  \
2544                                  Facebook         SOCIAL     4.1   
3943                                  Facebook         SOCIAL     4.1   
381                         WhatsApp Messenger  COMMUNICATION     4.4   
336                         WhatsApp Messenger  COMMUNICATION     4.4   
3904                        WhatsApp Messenger  COMMUNICATION     4.4   
2604                                 Instagram         SOCIAL     4.5   
2611                                 Instagram         SOCIAL     4.5   
2545                                 Instagram         SOCIAL     4.5   
3909                                 Instagram         SOCIAL     4.5   
382   Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
4104  Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
335   Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
1879                            Clash of Clans           GAME     4.6   
1704                            Clash of Clans           GAME     4.6   
1670                            Clash of Clans           GAME     4.6   

         Reviews                Size        Installs  Type Price  \
2544  78158306.0  Varies with device  1,000,000,000+  Free     0   
3943  78128208.0  Varies with device  1,000,000,000+  Free     0   
381   69119316.0  Varies with device  1,000,000,000+  Free     0   
336   69119316.0  Varies with device  1,000,000,000+  Free     0   
3904  69109672.0  Varies with device  1,000,000,000+  Free     0   
2604  66577446.0  Varies with device  1,000,000,000+  Free     0   
2611  66577313.0  Varies with device  1,000,000,000+  Free     0   
2545  66577313.0  Varies with device  1,000,000,000+  Free     0   
3909  66509917.0  Varies with device  1,000,000,000+  Free     0   
382   56646578.0  Varies with device  1,000,000,000+  Free     0   
4104  56642847.0  Varies with device  1,000,000,000+  Free     0   
335   56642847.0  Varies with device  1,000,000,000+  Free     0   
1879  44893888.0                 98M    100,000,000+  Free     0   
1704  44891723.0                 98M    100,000,000+  Free     0   
1670  44891723.0                 98M    100,000,000+  Free     0   

     Content Rating         Genres    Last Updated         Current Ver  \
2544           Teen         Social  August 3, 2018  Varies with device   
3943           Teen         Social  August 3, 2018  Varies with device   
381        Everyone  Communication  August 3, 2018  Varies with device   
336        Everyone  Communication  August 3, 2018  Varies with device   
3904       Everyone  Communication  August 3, 2018  Varies with device   
2604           Teen         Social   July 31, 2018  Varies with device   
2611           Teen         Social   July 31, 2018  Varies with device   
2545           Teen         Social   July 31, 2018  Varies with device   
3909           Teen         Social   July 31, 2018  Varies with device   
382        Everyone  Communication  August 1, 2018  Varies with device   
4104       Everyone  Communication  August 1, 2018  Varies with device   
335        Everyone  Communication  August 1, 2018  Varies with device   
1879   Everyone 10+       Strategy   July 15, 2018           10.322.16   
1704   Everyone 10+       Strategy   July 15, 2018           10.322.16   
1670   Everyone 10+       Strategy   July 15, 2018           10.322.16   

             Android Ver  
2544  Varies with device  
3943  Varies with device  
381   Varies with device  
336   Varies with device  
3904  Varies with device  
2604  Varies with device  
2611  Varies with device  
2545  Varies with device  
3909  Varies with device  
382   Varies with device  
4104  Varies with device  
335   Varies with device  
1879          4.1 and up  
1704          4.1 and up  
1670          4.1 and up

In [58]:
android.sort_values("Reviews",ascending=False, inplace=True)

In [59]:
android.head()

App       Category  Rating     Reviews  \
2544            Facebook         SOCIAL     4.1  78158306.0   
3943            Facebook         SOCIAL     4.1  78128208.0   
381   WhatsApp Messenger  COMMUNICATION     4.4  69119316.0   
336   WhatsApp Messenger  COMMUNICATION     4.4  69119316.0   
3904  WhatsApp Messenger  COMMUNICATION     4.4  69109672.0   

                    Size        Installs  Type Price Content Rating  \
2544  Varies with device  1,000,000,000+  Free     0           Teen   
3943  Varies with device  1,000,000,000+  Free     0           Teen   
381   Varies with device  1,000,000,000+  Free     0       Everyone   
336   Varies with device  1,000,000,000+  Free     0       Everyone   
3904  Varies with device  1,000,000,000+  Free     0       Everyone   

             Genres    Last Updated         Current Ver         Android Ver  
2544         Social  August 3, 2018  Varies with device  Varies with device  
3943         Social  August 3, 2018  Varies with device  Varies with device  
381   Communication  August 3, 2018  Varies with device  Varies with device  
336   Communication  August 3, 2018  Varies with device  Varies with device  
3904  Communication  August 3, 2018  Varies with device  Varies with device

In [60]:
android.drop_duplicates(["App"],keep="first",inplace=True) #bydefault it will keep 1st row, you can pass last as well

In [61]:
android

App           Category  \
2544                                           Facebook             SOCIAL   
381                                  WhatsApp Messenger      COMMUNICATION   
2604                                          Instagram             SOCIAL   
382            Messenger – Text and Video Chat for Free      COMMUNICATION   
1879                                     Clash of Clans               GAME   
...                                                 ...                ...   
9713                                          eP Finder  AUTO_AND_VEHICLES   
8495                                     HDWallpaper DK    PERSONALIZATION   
6277                                   Bi-Tank Ads Free               GAME   
8489                               DK Primrose for KLWP    PERSONALIZATION   
5086  Apps for SportsBєtting.ag - Bitcoin Welcome here!             SPORTS   

      Rating     Reviews                Size        Installs  Type  Price  \
2544     4.1  78158306.0  Varies with device  1,000,000,000+  Free      0   
381      4.4  69119316.0  Varies with device  1,000,000,000+  Free      0   
2604     4.5  66577446.0  Varies with device  1,000,000,000+  Free      0   
382      4.0  56646578.0  Varies with device  1,000,000,000+  Free      0   
1879     4.6  44893888.0                 98M    100,000,000+  Free      0   
...      ...         ...                 ...             ...   ...    ...   
9713     NaN         0.0                6.3M             10+  Free      0   
8495     NaN         0.0                6.6M             10+  Free      0   
6277     NaN         0.0  Varies with device              1+  Paid  $0.99   
8489     NaN         0.0                 16M              1+  Paid  $0.99   
5086     NaN         0.0               10.0M          1,000+  Free      0   

     Content Rating           Genres       Last Updated         Current Ver  \
2544           Teen           Social     August 3, 2018  Varies with device   
381        Everyone    Communication     August 3, 2018  Varies with device   
2604           Teen           Social      July 31, 2018  Varies with device   
382        Everyone    Communication     August 1, 2018  Varies with device   
1879   Everyone 10+         Strategy      July 15, 2018           10.322.16   
...             ...              ...                ...                 ...   
9713       Everyone  Auto & Vehicles      June 21, 2018                 1.0   
8495           Teen  Personalization    October 5, 2017                 1.0   
6277       Everyone           Arcade      July 31, 2017                 1.1   
8489       Everyone  Personalization   December 2, 2017     v2017.Dec.01.19   
5086       Everyone           Sports  December 11, 2017                 1.0   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up  
...                  ...  
9713          4.4 and up  
8495          4.0 and up  
6277          4.1 and up  
8489          4.4 and up  
5086          4.0 and up  

[9659 rows x 13 columns]

In [62]:
android.shape

(9659, 13)

**we have to work on free apps**

In [63]:
android["Type"]

2544    Free
381     Free
2604    Free
382     Free
1879    Free
        ... 
9713    Free
8495    Free
6277    Paid
8489    Paid
5086    Free
Name: Type, Length: 9659, dtype: object

In [64]:
android["Type"].unique()

array(['Free', 'Paid', nan], dtype=object)

In [65]:
android["Type"].value_counts() # not tell missing values

Free    8904
Paid     754
Name: Type, dtype: int64

In [66]:
android["Type"].value_counts(dropna=False) # not tell missing values

Free    8904
Paid     754
NaN        1
Name: Type, dtype: int64

In [67]:
android.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9659 entries, 2544 to 5086
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9659 non-null   object 
 1   Category        9659 non-null   object 
 2   Rating          8196 non-null   float64
 3   Reviews         9659 non-null   float64
 4   Size            9659 non-null   object 
 5   Installs        9659 non-null   object 
 6   Type            9658 non-null   object 
 7   Price           9659 non-null   object 
 8   Content Rating  9659 non-null   object 
 9   Genres          9659 non-null   object 
 10  Last Updated    9659 non-null   object 
 11  Current Ver     9651 non-null   object 
 12  Android Ver     9657 non-null   object 
dtypes: float64(2), object(11)
memory usage: 1.0+ MB


In [68]:
android["Type"]=="Free"

2544     True
381      True
2604     True
382      True
1879     True
        ...  
9713     True
8495     True
6277    False
8489    False
5086     True
Name: Type, Length: 9659, dtype: bool

In [69]:
free_app=android[android["Type"]=="Free"]
free_app

App           Category  \
2544                                           Facebook             SOCIAL   
381                                  WhatsApp Messenger      COMMUNICATION   
2604                                          Instagram             SOCIAL   
382            Messenger – Text and Video Chat for Free      COMMUNICATION   
1879                                     Clash of Clans               GAME   
...                                                 ...                ...   
8501                               DK Childcare Centers          LIFESTYLE   
2460                                    Drug Calculator            MEDICAL   
9713                                          eP Finder  AUTO_AND_VEHICLES   
8495                                     HDWallpaper DK    PERSONALIZATION   
5086  Apps for SportsBєtting.ag - Bitcoin Welcome here!             SPORTS   

      Rating     Reviews                Size        Installs  Type Price  \
2544     4.1  78158306.0  Varies with device  1,000,000,000+  Free     0   
381      4.4  69119316.0  Varies with device  1,000,000,000+  Free     0   
2604     4.5  66577446.0  Varies with device  1,000,000,000+  Free     0   
382      4.0  56646578.0  Varies with device  1,000,000,000+  Free     0   
1879     4.6  44893888.0                 98M    100,000,000+  Free     0   
...      ...         ...                 ...             ...   ...   ...   
8501     NaN         0.0                 13M            500+  Free     0   
2460     NaN         0.0                4.6M             10+  Free     0   
9713     NaN         0.0                6.3M             10+  Free     0   
8495     NaN         0.0                6.6M             10+  Free     0   
5086     NaN         0.0               10.0M          1,000+  Free     0   

     Content Rating           Genres       Last Updated         Current Ver  \
2544           Teen           Social     August 3, 2018  Varies with device   
381        Everyone    Communication     August 3, 2018  Varies with device   
2604           Teen           Social      July 31, 2018  Varies with device   
382        Everyone    Communication     August 1, 2018  Varies with device   
1879   Everyone 10+         Strategy      July 15, 2018           10.322.16   
...             ...              ...                ...                 ...   
8501       Everyone        Lifestyle      April 2, 2018               1.0.2   
2460       Everyone          Medical      July 27, 2018                 1.0   
9713       Everyone  Auto & Vehicles      June 21, 2018                 1.0   
8495           Teen  Personalization    October 5, 2017                 1.0   
5086       Everyone           Sports  December 11, 2017                 1.0   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up  
...                  ...  
8501          4.1 and up  
2460          4.4 and up  
9713          4.4 and up  
8495          4.0 and up  
5086          4.0 and up  

[8904 rows x 13 columns]

In [70]:
free_app.shape

(8904, 13)

**we just need english app only**

In [71]:
free_app

App           Category  \
2544                                           Facebook             SOCIAL   
381                                  WhatsApp Messenger      COMMUNICATION   
2604                                          Instagram             SOCIAL   
382            Messenger – Text and Video Chat for Free      COMMUNICATION   
1879                                     Clash of Clans               GAME   
...                                                 ...                ...   
8501                               DK Childcare Centers          LIFESTYLE   
2460                                    Drug Calculator            MEDICAL   
9713                                          eP Finder  AUTO_AND_VEHICLES   
8495                                     HDWallpaper DK    PERSONALIZATION   
5086  Apps for SportsBєtting.ag - Bitcoin Welcome here!             SPORTS   

      Rating     Reviews                Size        Installs  Type Price  \
2544     4.1  78158306.0  Varies with device  1,000,000,000+  Free     0   
381      4.4  69119316.0  Varies with device  1,000,000,000+  Free     0   
2604     4.5  66577446.0  Varies with device  1,000,000,000+  Free     0   
382      4.0  56646578.0  Varies with device  1,000,000,000+  Free     0   
1879     4.6  44893888.0                 98M    100,000,000+  Free     0   
...      ...         ...                 ...             ...   ...   ...   
8501     NaN         0.0                 13M            500+  Free     0   
2460     NaN         0.0                4.6M             10+  Free     0   
9713     NaN         0.0                6.3M             10+  Free     0   
8495     NaN         0.0                6.6M             10+  Free     0   
5086     NaN         0.0               10.0M          1,000+  Free     0   

     Content Rating           Genres       Last Updated         Current Ver  \
2544           Teen           Social     August 3, 2018  Varies with device   
381        Everyone    Communication     August 3, 2018  Varies with device   
2604           Teen           Social      July 31, 2018  Varies with device   
382        Everyone    Communication     August 1, 2018  Varies with device   
1879   Everyone 10+         Strategy      July 15, 2018           10.322.16   
...             ...              ...                ...                 ...   
8501       Everyone        Lifestyle      April 2, 2018               1.0.2   
2460       Everyone          Medical      July 27, 2018                 1.0   
9713       Everyone  Auto & Vehicles      June 21, 2018                 1.0   
8495           Teen  Personalization    October 5, 2017                 1.0   
5086       Everyone           Sports  December 11, 2017                 1.0   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up  
...                  ...  
8501          4.1 and up  
2460          4.4 and up  
9713          4.4 and up  
8495          4.0 and up  
5086          4.0 and up  

[8904 rows x 13 columns]

In [72]:
free_app["App"].tail()

8501                                 DK Childcare Centers
2460                                      Drug Calculator
9713                                            eP Finder
8495                                       HDWallpaper DK
5086    Apps for SportsBєtting.ag - Bitcoin Welcome here!
Name: App, dtype: object

In [73]:
free_app["App"].head()

2544                                    Facebook
381                           WhatsApp Messenger
2604                                   Instagram
382     Messenger – Text and Video Chat for Free
1879                              Clash of Clans
Name: App, dtype: object

In [74]:
free_app["App"]

2544                                             Facebook
381                                    WhatsApp Messenger
2604                                            Instagram
382              Messenger – Text and Video Chat for Free
1879                                       Clash of Clans
                              ...                        
8501                                 DK Childcare Centers
2460                                      Drug Calculator
9713                                            eP Finder
8495                                       HDWallpaper DK
5086    Apps for SportsBєtting.ag - Bitcoin Welcome here!
Name: App, Length: 8904, dtype: object

In [75]:
free_app["App"].sort_values().head(50)

8884                  "i DT" Fútbol. Todos Somos Técnicos.
8532                         +Download 4 Instagram Twitter
324                             - Free Comics - Comic Apps
4541                                                    .R
4636                                                /u/app
5940                                                058.ba
10252                                       1. FC Köln App
1407                                 10 Best Foods for You
8483                         10 Minutes a Day Times Tables
9401                                  100 Doors of Revenge
4030                                       100+ C Programs
8703       100000+ Messages - DP, Status, Jokes & GIF 2018
4026                            101 C Programming Problems
220      104 Looking for a job - looking for a job, loo...
2694                                                  11st
5344     12 Step Meditations & Sober Prayers AA NA AL-ANON
2483                                         14thStreetV

In [76]:
free_app

App           Category  \
2544                                           Facebook             SOCIAL   
381                                  WhatsApp Messenger      COMMUNICATION   
2604                                          Instagram             SOCIAL   
382            Messenger – Text and Video Chat for Free      COMMUNICATION   
1879                                     Clash of Clans               GAME   
...                                                 ...                ...   
8501                               DK Childcare Centers          LIFESTYLE   
2460                                    Drug Calculator            MEDICAL   
9713                                          eP Finder  AUTO_AND_VEHICLES   
8495                                     HDWallpaper DK    PERSONALIZATION   
5086  Apps for SportsBєtting.ag - Bitcoin Welcome here!             SPORTS   

      Rating     Reviews                Size        Installs  Type Price  \
2544     4.1  78158306.0  Varies with device  1,000,000,000+  Free     0   
381      4.4  69119316.0  Varies with device  1,000,000,000+  Free     0   
2604     4.5  66577446.0  Varies with device  1,000,000,000+  Free     0   
382      4.0  56646578.0  Varies with device  1,000,000,000+  Free     0   
1879     4.6  44893888.0                 98M    100,000,000+  Free     0   
...      ...         ...                 ...             ...   ...   ...   
8501     NaN         0.0                 13M            500+  Free     0   
2460     NaN         0.0                4.6M             10+  Free     0   
9713     NaN         0.0                6.3M             10+  Free     0   
8495     NaN         0.0                6.6M             10+  Free     0   
5086     NaN         0.0               10.0M          1,000+  Free     0   

     Content Rating           Genres       Last Updated         Current Ver  \
2544           Teen           Social     August 3, 2018  Varies with device   
381        Everyone    Communication     August 3, 2018  Varies with device   
2604           Teen           Social      July 31, 2018  Varies with device   
382        Everyone    Communication     August 1, 2018  Varies with device   
1879   Everyone 10+         Strategy      July 15, 2018           10.322.16   
...             ...              ...                ...                 ...   
8501       Everyone        Lifestyle      April 2, 2018               1.0.2   
2460       Everyone          Medical      July 27, 2018                 1.0   
9713       Everyone  Auto & Vehicles      June 21, 2018                 1.0   
8495           Teen  Personalization    October 5, 2017                 1.0   
5086       Everyone           Sports  December 11, 2017                 1.0   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up  
...                  ...  
8501          4.1 and up  
2460          4.4 and up  
9713          4.4 and up  
8495          4.0 and up  
5086          4.0 and up  

[8904 rows x 13 columns]

**0-127 ASCII** >>128 character
* our data is in unicode

In [77]:
def clean_eng(app):
    print(app)
    for i in app:
        if ord(i)<128:
            return True
        else:
            return False
        

In [78]:
clean_eng("Faceboodajdidjaisdjasioadj<(a#k >()")

Faceboodajdidjaisdjasioadj<(a#k >()


True

In [79]:
clean_eng("Faceboodajdidjaisdjasioadj<(a#k >()😍")

Faceboodajdidjaisdjasioadj<(a#k >()😍


True

In [80]:
ord("😍")

128525

In [81]:
ord(")")

41

In [82]:
def clean_eng(app):
    #print(app)
    non_eng=0
    for i in app:
        #print(i)
        if ord(i)>128:
            non_eng += 1
    if non_eng>3:
        return False
    else:
        return True
        

In [83]:
clean_eng("Faceboodajdidjaisdjasioadj<(a#k >()😍")

True

In [86]:
clean_eng("Faceboodajdidjaisdjasioadj<(a#k >()😍😍😍😍😍😍")

False

**such a function that is not availbe in python before we can use keyword apply on the function**

**it will return true and false**

In [87]:
free_app.App.apply(clean_eng)

2544    True
381     True
2604    True
382     True
1879    True
        ... 
8501    True
2460    True
9713    True
8495    True
5086    True
Name: App, Length: 8904, dtype: bool

In [91]:
free_app.App.apply(clean_eng).all() #means that there is some false values also exit

False

In [92]:
free_app["App"].apply(clean_eng).sum()  # maeans these values are true , it just count true value 

8863

**we can do this with value counts as well**

In [93]:
free_app["App"].apply(clean_eng).value_counts()  # maeans these values are true 

True     8863
False      41
Name: App, dtype: int64

In [94]:
is_eng_app=free_app["App"].apply(clean_eng)# i can use it as boolen index to find that is it true or false
is_eng_app

2544    True
381     True
2604    True
382     True
1879    True
        ... 
8501    True
2460    True
9713    True
8495    True
5086    True
Name: App, Length: 8904, dtype: bool

In [95]:
final_app=free_app[is_eng_app]# boolean index always return true values
final_app

App           Category  \
2544                                           Facebook             SOCIAL   
381                                  WhatsApp Messenger      COMMUNICATION   
2604                                          Instagram             SOCIAL   
382            Messenger – Text and Video Chat for Free      COMMUNICATION   
1879                                     Clash of Clans               GAME   
...                                                 ...                ...   
8501                               DK Childcare Centers          LIFESTYLE   
2460                                    Drug Calculator            MEDICAL   
9713                                          eP Finder  AUTO_AND_VEHICLES   
8495                                     HDWallpaper DK    PERSONALIZATION   
5086  Apps for SportsBєtting.ag - Bitcoin Welcome here!             SPORTS   

      Rating     Reviews                Size        Installs  Type Price  \
2544     4.1  78158306.0  Varies with device  1,000,000,000+  Free     0   
381      4.4  69119316.0  Varies with device  1,000,000,000+  Free     0   
2604     4.5  66577446.0  Varies with device  1,000,000,000+  Free     0   
382      4.0  56646578.0  Varies with device  1,000,000,000+  Free     0   
1879     4.6  44893888.0                 98M    100,000,000+  Free     0   
...      ...         ...                 ...             ...   ...   ...   
8501     NaN         0.0                 13M            500+  Free     0   
2460     NaN         0.0                4.6M             10+  Free     0   
9713     NaN         0.0                6.3M             10+  Free     0   
8495     NaN         0.0                6.6M             10+  Free     0   
5086     NaN         0.0               10.0M          1,000+  Free     0   

     Content Rating           Genres       Last Updated         Current Ver  \
2544           Teen           Social     August 3, 2018  Varies with device   
381        Everyone    Communication     August 3, 2018  Varies with device   
2604           Teen           Social      July 31, 2018  Varies with device   
382        Everyone    Communication     August 1, 2018  Varies with device   
1879   Everyone 10+         Strategy      July 15, 2018           10.322.16   
...             ...              ...                ...                 ...   
8501       Everyone        Lifestyle      April 2, 2018               1.0.2   
2460       Everyone          Medical      July 27, 2018                 1.0   
9713       Everyone  Auto & Vehicles      June 21, 2018                 1.0   
8495           Teen  Personalization    October 5, 2017                 1.0   
5086       Everyone           Sports  December 11, 2017                 1.0   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up  
...                  ...  
8501          4.1 and up  
2460          4.4 and up  
9713          4.4 and up  
8495          4.0 and up  
5086          4.0 and up  

[8863 rows x 13 columns]

In [96]:
final_app.head()

App       Category  Rating  \
2544                                  Facebook         SOCIAL     4.1   
381                         WhatsApp Messenger  COMMUNICATION     4.4   
2604                                 Instagram         SOCIAL     4.5   
382   Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
1879                            Clash of Clans           GAME     4.6   

         Reviews                Size        Installs  Type Price  \
2544  78158306.0  Varies with device  1,000,000,000+  Free     0   
381   69119316.0  Varies with device  1,000,000,000+  Free     0   
2604  66577446.0  Varies with device  1,000,000,000+  Free     0   
382   56646578.0  Varies with device  1,000,000,000+  Free     0   
1879  44893888.0                 98M    100,000,000+  Free     0   

     Content Rating         Genres    Last Updated         Current Ver  \
2544           Teen         Social  August 3, 2018  Varies with device   
381        Everyone  Communication  August 3, 2018  Varies with device   
2604           Teen         Social   July 31, 2018  Varies with device   
382        Everyone  Communication  August 1, 2018  Varies with device   
1879   Everyone 10+       Strategy   July 15, 2018           10.322.16   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up

In [97]:
final_app.shape

(8863, 13)

# Analysis part 

**we can find free app by view the rating,maximum review, number of installation in device colums**

**review mean that you can write any thing about app but rating and no. of installation are two main factor to find free app**

**most important is no. of installation**

**we can drop Type and Price colum part because we did not need more**


In [98]:
final_app.drop(["Type","Price"],axis=1,inplace=True)

C:\Users\junaid.8646059\Junaid\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [99]:
final_app

App           Category  \
2544                                           Facebook             SOCIAL   
381                                  WhatsApp Messenger      COMMUNICATION   
2604                                          Instagram             SOCIAL   
382            Messenger – Text and Video Chat for Free      COMMUNICATION   
1879                                     Clash of Clans               GAME   
...                                                 ...                ...   
8501                               DK Childcare Centers          LIFESTYLE   
2460                                    Drug Calculator            MEDICAL   
9713                                          eP Finder  AUTO_AND_VEHICLES   
8495                                     HDWallpaper DK    PERSONALIZATION   
5086  Apps for SportsBєtting.ag - Bitcoin Welcome here!             SPORTS   

      Rating     Reviews                Size        Installs Content Rating  \
2544     4.1  78158306.0  Varies with device  1,000,000,000+           Teen   
381      4.4  69119316.0  Varies with device  1,000,000,000+       Everyone   
2604     4.5  66577446.0  Varies with device  1,000,000,000+           Teen   
382      4.0  56646578.0  Varies with device  1,000,000,000+       Everyone   
1879     4.6  44893888.0                 98M    100,000,000+   Everyone 10+   
...      ...         ...                 ...             ...            ...   
8501     NaN         0.0                 13M            500+       Everyone   
2460     NaN         0.0                4.6M             10+       Everyone   
9713     NaN         0.0                6.3M             10+       Everyone   
8495     NaN         0.0                6.6M             10+           Teen   
5086     NaN         0.0               10.0M          1,000+       Everyone   

               Genres       Last Updated         Current Ver  \
2544           Social     August 3, 2018  Varies with device   
381     Communication     August 3, 2018  Varies with device   
2604           Social      July 31, 2018  Varies with device   
382     Communication     August 1, 2018  Varies with device   
1879         Strategy      July 15, 2018           10.322.16   
...               ...                ...                 ...   
8501        Lifestyle      April 2, 2018               1.0.2   
2460          Medical      July 27, 2018                 1.0   
9713  Auto & Vehicles      June 21, 2018                 1.0   
8495  Personalization    October 5, 2017                 1.0   
5086           Sports  December 11, 2017                 1.0   

             Android Ver  
2544  Varies with device  
381   Varies with device  
2604  Varies with device  
382   Varies with device  
1879          4.1 and up  
...                  ...  
8501          4.1 and up  
2460          4.4 and up  
9713          4.4 and up  
8495          4.0 and up  
5086          4.0 and up  

[8863 rows x 11 columns]

In [100]:
final_app.sort_index().head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   

    Reviews  Size     Installs Content Rating                   Genres  \
0     159.0   19M      10,000+       Everyone             Art & Design   
2   87510.0  8.7M   5,000,000+       Everyone             Art & Design   
3  215644.0   25M  50,000,000+           Teen             Art & Design   
4     967.0  2.8M     100,000+       Everyone  Art & Design;Creativity   
5     167.0  5.6M      50,000+       Everyone             Art & Design   

      Last Updated         Current Ver   Android Ver  
0  January 7, 2018               1.0.0  4.0.3 and up  
2   August 1, 2018               1.2.4  4.0.3 and up  
3     June 8, 2018  Varies with device    4.2 and up  
4    June 20, 2018                 1.1    4.4 and up  
5   March 26, 2017                 1.0    2.3 and up

In [101]:
final_app.head()

App       Category  Rating  \
2544                                  Facebook         SOCIAL     4.1   
381                         WhatsApp Messenger  COMMUNICATION     4.4   
2604                                 Instagram         SOCIAL     4.5   
382   Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
1879                            Clash of Clans           GAME     4.6   

         Reviews                Size        Installs Content Rating  \
2544  78158306.0  Varies with device  1,000,000,000+           Teen   
381   69119316.0  Varies with device  1,000,000,000+       Everyone   
2604  66577446.0  Varies with device  1,000,000,000+           Teen   
382   56646578.0  Varies with device  1,000,000,000+       Everyone   
1879  44893888.0                 98M    100,000,000+   Everyone 10+   

             Genres    Last Updated         Current Ver         Android Ver  
2544         Social  August 3, 2018  Varies with device  Varies with device  
381   Communication  August 3, 2018  Varies with device  Varies with device  
2604         Social   July 31, 2018  Varies with device  Varies with device  
382   Communication  August 1, 2018  Varies with device  Varies with device  
1879       Strategy   July 15, 2018           10.322.16          4.1 and up

**we now need to find avg of installation with respect to category 

**first we need to remove + before convert into float or int**

**we use str because it is string ,, this is happend only in pandas series**

In [102]:
final_app["Installs"]=final_app["Installs"].str.replace(",","")\
.str.replace("+","").astype(int)

<ipython-input-102-554fc0f8b8fa>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_app["Installs"]=final_app["Installs"].str.replace(",","")\


In [103]:
#just for check the sum , no aim of this line
final_app.Installs.sum()# means our data is now in integer and \ break the line if our code line is long

75251051335

**now we need to count the social app** 

**for this we can use value count**

In [104]:
final_app["Category"].value_counts()

FAMILY                 1676
GAME                    861
TOOLS                   750
BUSINESS                407
LIFESTYLE               346
PRODUCTIVITY            345
FINANCE                 328
MEDICAL                 313
SPORTS                  301
PERSONALIZATION         294
COMMUNICATION           287
HEALTH_AND_FITNESS      273
PHOTOGRAPHY             261
NEWS_AND_MAGAZINES      248
SOCIAL                  236
TRAVEL_AND_LOCAL        207
SHOPPING                199
BOOKS_AND_REFERENCE     190
DATING                  165
VIDEO_PLAYERS           159
MAPS_AND_NAVIGATION     124
FOOD_AND_DRINK          110
EDUCATION               103
ENTERTAINMENT            85
LIBRARIES_AND_DEMO       83
AUTO_AND_VEHICLES        82
HOUSE_AND_HOME           73
WEATHER                  71
EVENTS                   63
PARENTING                58
ART_AND_DESIGN           57
COMICS                   55
BEAUTY                   53
Name: Category, dtype: int64

In [105]:
final_app["Category"].value_counts(ascending=True)

BEAUTY                   53
COMICS                   55
ART_AND_DESIGN           57
PARENTING                58
EVENTS                   63
WEATHER                  71
HOUSE_AND_HOME           73
AUTO_AND_VEHICLES        82
LIBRARIES_AND_DEMO       83
ENTERTAINMENT            85
EDUCATION               103
FOOD_AND_DRINK          110
MAPS_AND_NAVIGATION     124
VIDEO_PLAYERS           159
DATING                  165
BOOKS_AND_REFERENCE     190
SHOPPING                199
TRAVEL_AND_LOCAL        207
SOCIAL                  236
NEWS_AND_MAGAZINES      248
PHOTOGRAPHY             261
HEALTH_AND_FITNESS      273
COMMUNICATION           287
PERSONALIZATION         294
SPORTS                  301
MEDICAL                 313
FINANCE                 328
PRODUCTIVITY            345
LIFESTYLE               346
BUSINESS                407
TOOLS                   750
GAME                    861
FAMILY                 1676
Name: Category, dtype: int64

**FAMILY 1676 does not means that it is install most time so we need to find the mean**

In [106]:
cat=final_app["Category"].unique()
cat

array(['SOCIAL', 'COMMUNICATION', 'GAME', 'TOOLS', 'VIDEO_PLAYERS',
       'NEWS_AND_MAGAZINES', 'PHOTOGRAPHY', 'FAMILY', 'TRAVEL_AND_LOCAL',
       'PERSONALIZATION', 'MAPS_AND_NAVIGATION', 'SHOPPING',
       'ENTERTAINMENT', 'PRODUCTIVITY', 'HEALTH_AND_FITNESS', 'SPORTS',
       'BOOKS_AND_REFERENCE', 'LIFESTYLE', 'WEATHER', 'FINANCE',
       'BUSINESS', 'EDUCATION', 'FOOD_AND_DRINK', 'COMICS', 'PARENTING',
       'DATING', 'HOUSE_AND_HOME', 'LIBRARIES_AND_DEMO', 'ART_AND_DESIGN',
       'AUTO_AND_VEHICLES', 'MEDICAL', 'BEAUTY', 'EVENTS'], dtype=object)

In [107]:
len(final_app["Category"].unique())

33

In [108]:
for i in cat[:3]:
    print(i)

SOCIAL
COMMUNICATION
GAME


In [109]:
cat_mean={}
for i in cat[:3]:
    df=final_app[final_app["Category"]==i] # this boolean indexing is on data frame
    print(df)

                 App Category  Rating     Reviews                Size  \
2544        Facebook   SOCIAL     4.1  78158306.0  Varies with device   
2604       Instagram   SOCIAL     4.5  66577446.0  Varies with device   
2610        Snapchat   SOCIAL     4.0  17015352.0  Varies with device   
2546   Facebook Lite   SOCIAL     4.3   8606259.0  Varies with device   
6373              VK   SOCIAL     3.8   5793284.0  Varies with device   
...              ...      ...     ...         ...                 ...   
9330     EG Way Life   SOCIAL     NaN         0.0                1.1M   
9137        quran-DZ   SOCIAL     NaN         0.0                6.2M   
10101      Amleen Ey   SOCIAL     NaN         0.0                 19M   
8579         Otto DM   SOCIAL     NaN         0.0                2.4M   
6257      BH Connect   SOCIAL     NaN         0.0                 32M   

         Installs Content Rating  Genres    Last Updated         Current Ver  \
2544   1000000000           Teen  Social  A

In [110]:
cat_mean={}
for i in cat[:3]:
    df=final_app[final_app["Category"]==i]
    mean=df["Installs"].mean()
    cat_mean[i]=mean

In [111]:
cat_mean

{'SOCIAL': 23253652.127118643,
 'COMMUNICATION': 38456119.167247385,
 'GAME': 15594505.74912892}

In [ ]:
#final_app[final_app["Category"]=="SOCIAL"]

In [112]:
cat_mean={}
for i in cat[:]:
    df=final_app[final_app["Category"]==i]
    mean=df["Installs"].mean()
    cat_mean[i]=mean

In [113]:
cat_mean

{'SOCIAL': 23253652.127118643,
 'COMMUNICATION': 38456119.167247385,
 'GAME': 15594505.74912892,
 'TOOLS': 10801391.298666667,
 'VIDEO_PLAYERS': 24727872.452830188,
 'NEWS_AND_MAGAZINES': 9549178.467741935,
 'PHOTOGRAPHY': 17840110.40229885,
 'FAMILY': 3701608.406921241,
 'TRAVEL_AND_LOCAL': 13984077.710144928,
 'PERSONALIZATION': 5201482.6122448975,
 'MAPS_AND_NAVIGATION': 4056941.7741935486,
 'SHOPPING': 7036877.311557789,
 'ENTERTAINMENT': 11640705.88235294,
 'PRODUCTIVITY': 16787331.344927534,
 'HEALTH_AND_FITNESS': 4188821.9853479853,
 'SPORTS': 3638640.1428571427,
 'BOOKS_AND_REFERENCE': 8767811.894736841,
 'LIFESTYLE': 1437816.2687861272,
 'WEATHER': 5074486.197183099,
 'FINANCE': 1387692.475609756,
 'BUSINESS': 1712290.1474201474,
 'EDUCATION': 1833495.145631068,
 'FOOD_AND_DRINK': 1924897.7363636363,
 'COMICS': 817657.2727272727,
 'PARENTING': 542603.6206896552,
 'DATING': 854028.8303030303,
 'HOUSE_AND_HOME': 1331540.5616438356,
 'LIBRARIES_AND_DEMO': 638503.734939759,
 'ART_

**above method is defficult**

In [114]:
pop_app=final_app.groupby("Category")["Installs"].mean()
pop_app
#this line will make data frame and then go to installs and calculate mean

Category
ART_AND_DESIGN         1.986335e+06
AUTO_AND_VEHICLES      6.473178e+05
BEAUTY                 5.131519e+05
BOOKS_AND_REFERENCE    8.767812e+06
BUSINESS               1.712290e+06
COMICS                 8.176573e+05
COMMUNICATION          3.845612e+07
DATING                 8.540288e+05
EDUCATION              1.833495e+06
ENTERTAINMENT          1.164071e+07
EVENTS                 2.535422e+05
FAMILY                 3.701608e+06
FINANCE                1.387692e+06
FOOD_AND_DRINK         1.924898e+06
GAME                   1.559451e+07
HEALTH_AND_FITNESS     4.188822e+06
HOUSE_AND_HOME         1.331541e+06
LIBRARIES_AND_DEMO     6.385037e+05
LIFESTYLE              1.437816e+06
MAPS_AND_NAVIGATION    4.056942e+06
MEDICAL                1.205506e+05
NEWS_AND_MAGAZINES     9.549178e+06
PARENTING              5.426036e+05
PERSONALIZATION        5.201483e+06
PHOTOGRAPHY            1.784011e+07
PRODUCTIVITY           1.678733e+07
SHOPPING               7.036877e+06
SOCIAL             

In [115]:
pop_app.sort_values(ascending=False).head(10)
#most popular app is used as english and free app is communication appliction

Category
COMMUNICATION         3.845612e+07
VIDEO_PLAYERS         2.472787e+07
SOCIAL                2.325365e+07
PHOTOGRAPHY           1.784011e+07
PRODUCTIVITY          1.678733e+07
GAME                  1.559451e+07
TRAVEL_AND_LOCAL      1.398408e+07
ENTERTAINMENT         1.164071e+07
TOOLS                 1.080139e+07
NEWS_AND_MAGAZINES    9.549178e+06
Name: Installs, dtype: float64

# now i need to check the join in this catagory 

In [116]:
com_app=final_app[final_app["Category"]=="COMMUNICATION"]
com_app[["App","Installs"]].sort_values("Installs",ascending=False).head(50)
#we can not bear these joins because these are so strong commpany,
#there is  no soubt tht communication is most used but we go with some other category
#2nd is we should remove some higher companies that effect our means 

App    Installs
381                                  WhatsApp Messenger  1000000000
468                       Skype - free IM & video calls  1000000000
382            Messenger – Text and Video Chat for Free  1000000000
451                                               Gmail  1000000000
411                        Google Chrome: Fast & Secure  1000000000
464                                            Hangouts  1000000000
474                         LINE: Free Calls & Messages   500000000
4676                                    Viber Messenger   500000000
383                       imo free video calls and chat   500000000
4039              Google Duo - High Quality Video Calls   500000000
420         UC Browser - Fast Download Private & Secure   500000000
369                                    Android Messages   100000000
346                        imo beta free calls and text   100000000
4106              Messenger Lite: Free Calls & Messages   100000000
393                                                 Who   100000000
4279                                                Kik   100000000
4422                     Opera Browser: Fast and Secure   100000000
4302                       KakaoTalk: Free Calls & Text   100000000
395          GO SMS Pro - Messenger, Free Themes, Emoji   100000000
412                      Firefox Browser fast & private   100000000
4592                                           Telegram   100000000
4633        UC Browser Mini -Tiny Fast Private & Secure   100000000
4800                        Yahoo Mail – Stay Organized   100000000
6449                        BBM - Free Calls & Messages   100000000
4421                      Opera Mini - fast web browser   100000000
4712                                             WeChat   100000000
4602  Truecaller: Caller ID, SMS spam blocking & Dialer   100000000
6297                                  Zalo – Video Call    50000000
389                           free video calls and chat    50000000
459                                 Mail.Ru - Email App    50000000
348                                            Contacts    50000000
3892                                               Azar    50000000
418          Dolphin Browser - Fast, Private & Adblock🐬    50000000
4949  CM Browser - Ad Blocker , Fast Download , Privacy    50000000
4727                                  WhatsApp Business    10000000
375                               AT&T Visual Voicemail    10000000
4739  WhatsCall Free Global Phone Call App & Cheap C...    10000000
435                            Hiya - Caller ID & Block    10000000
437                       Mr. Number-Block calls & spam    10000000
5594                                       Google Voice    10000000
4369                                  Free WiFi Connect    10000000
4942   Free Adblocker Browser - Adblock & Popup Blocker    10000000
4609                             Portable Wi-Fi hotspot    10000000
4239        Talkatone: Free Texts, Calls & Phone Number    10000000
355                            Cricket Visual Voicemail    10000000
337                                   Messenger for SMS    10000000
429                        ExDialer - Dialer & Contacts    10000000
458                                         GO Notifier    10000000
9856                      Hangouts Dialer - Call Phones    10000000
4948                        Adblock Browser for Android    10000000

In [117]:
com_app[com_app["Installs"]<100000000]["Installs"].mean()
#so we filter the app and calculate the mean of installl colums again 
#3.845612e+07-3603485.3884615386

3603485.3884615386

# Summary

* most popular app is used as english and free app is communication appliction
* but there are many strong company inside communication, we can not bear them
* so check an other field , there is book and reference also
* we can design a book with some new features, with quiz and reward
* data science is creative a new think, idea,and new way in the business
* rating is no good way to make decision